quarkus-messaging-kafka (Kotlin) – Error MetadataExtensions: Provider JvmMetadataExtensions not found in 3.26.4 and lower (works in <= 3.20.2 LTS)

### Describe the bug

When using Kotlin with quarkus-messaging-kafka, sending a message to Kafka fails on Quarkus 3.26.4 with an error related to the Kotlin reflection / Jackson Kotlin module.
This issue does not occur on 3.20.2 LTS, so it seems to be a regression introduced between these versions.


```
@ApplicationScoped
class AgentService {

    @Inject
    @Channel("agents")
    lateinit var resources: MutinyEmitter<OnEventStoreMessage>

    fun onJoined(event: OnAgentJoined): Uni<Void> {
        return resources.send(
            OnEventStoreAgentConnected(event.agent.id.serial, event.since)
        )
    }
}
```



### Expected behavior

SRMSG18211: Message org.eclipse.microprofile.reactive.messaging.Message$5@240e4074 from channel 'agents' sent successfully to Kafka topic-partition 'xxxxx-prd-hexa-agents-0', with offset 1400

### Actual behavior

Exception

Exception 
```
025-09-18 14:33:36,161 DEBUG [io.sma.rea.mes.kafka] (vert.x-eventloop-thread-15) SRMSG18209: Sending message org.eclipse.microprofile.reactive.messaging.Message$5@14961151 from channel 'agents' to Kafka topic 'xxxxxx-prd-hexa-agents'
2025-09-18 14:35:08,128 ERROR [io.sma.rea.mes.kafka] (smallrye-kafka-producer-thread-2) SRMSG18212: Message org.eclipse.microprofile.reactive.messaging.Message$5@37682c4e from channel 'agents' was not sent to Kafka topic 'xxxxxx-prd-hexa-agents' - nacking message: java.lang.IllegalStateException: Retries exhausted : 10 attempts against 1758206126277/1758206124096 expiration
	at io.smallrye.mutiny.helpers.ExponentialBackoff$2.lambda$apply$0(ExponentialBackoff.java:120)
	at io.smallrye.context.impl.wrappers.SlowContextualFunction.apply(SlowContextualFunction.java:21)
	at io.smallrye.mutiny.groups.MultiOnItem.lambda$transformToUni$6(MultiOnItem.java:278)
	at io.smallrye.mutiny.operators.multi.MultiConcatMapOp$MainSubscriber.onItem(MultiConcatMapOp.java:117)
	at io.smallrye.mutiny.subscription.MultiSubscriber.onNext(MultiSubscriber.java:61)
	at io.smallrye.mutiny.operators.multi.processors.UnicastProcessor.drainWithDownstream(UnicastProcessor.java:108)
	at io.smallrye.mutiny.operators.multi.processors.UnicastProcessor.drain(UnicastProcessor.java:139)
	at io.smallrye.mutiny.operators.multi.processors.UnicastProcessor.onNext(UnicastProcessor.java:205)
	at io.smallrye.mutiny.operators.multi.processors.SerializedProcessor.onNext(SerializedProcessor.java:104)
	at io.smallrye.mutiny.subscription.SerializedSubscriber.onItem(SerializedSubscriber.java:74)
	at io.smallrye.mutiny.subscription.MultiSubscriber.onNext(MultiSubscriber.java:61)
	at io.smallrye.mutiny.operators.multi.MultiRetryWhenOp$RetryWhenOperator.onFailure(MultiRetryWhenOp.java:127)
	at io.smallrye.mutiny.subscription.MultiSubscriber.onError(MultiSubscriber.java:73)
	at io.smallrye.mutiny.converters.uni.UniToMultiPublisher$UniToMultiSubscription.onFailure(UniToMultiPublisher.java:104)
	at io.smallrye.mutiny.operators.uni.UniOperatorProcessor.onFailure(UniOperatorProcessor.java:55)
	at io.smallrye.mutiny.operators.uni.UniOperatorProcessor.onFailure(UniOperatorProcessor.java:55)
	at io.smallrye.mutiny.operators.uni.builders.DefaultUniEmitter.fail(DefaultUniEmitter.java:53)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateWithEmitter.subscribe(UniCreateWithEmitter.java:26)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:35)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.performInnerSubscription(UniOnItemTransformToUni.java:81)
	at io.smallrye.mutiny.operators.uni.UniOnItemTransformToUni$UniOnItemTransformToUniProcessor.onItem(UniOnItemTransformToUni.java:57)
	at io.smallrye.mutiny.operators.uni.UniOperatorProcessor.onItem(UniOperatorProcessor.java:47)
	at io.smallrye.mutiny.operators.uni.UniMemoizeOp.forwardTo(UniMemoizeOp.java:150)
	at io.smallrye.mutiny.operators.uni.UniMemoizeOp.subscribe(UniMemoizeOp.java:84)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:35)
	at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.lambda$subscribe$0(UniRunSubscribeOn.java:27)
	at java.base@21.0.5/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base@21.0.5/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base@21.0.5/java.lang.Thread.runWith(Thread.java:1596)
	at java.base@21.0.5/java.lang.Thread.run(Thread.java:1583)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:896)
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:872)
Caused by: java.util.ServiceConfigurationError: kotlin.reflect.jvm.internal.impl.km.internal.extensions.MetadataExtensions: Provider kotlin.reflect.jvm.internal.impl.km.jvm.internal.JvmMetadataExtensions not found
	at java.base@21.0.5/java.util.ServiceLoader.fail(ServiceLoader.java:593)
	at java.base@21.0.5/java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1219)
	at java.base@21.0.5/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1228)
	at java.base@21.0.5/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1273)
	at java.base@21.0.5/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1309)
	at java.base@21.0.5/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1393)
	at kotlin.collections.CollectionsKt___CollectionsKt.toCollection(_Collections.kt:1309)
	at kotlin.collections.CollectionsKt___CollectionsKt.toMutableList(_Collections.kt:1342)
	at kotlin.collections.CollectionsKt___CollectionsKt.toList(_Collections.kt:1333)
	at kotlin.reflect.jvm.internal.impl.km.internal.extensions.MetadataExtensions$Companion.INSTANCES_delegate$lambda$1(MetadataExtensions.kt:93)
	at kotlin.reflect.jvm.internal.impl.km.internal.extensions.MetadataExtensions$Companion.accessor$MetadataExtensions$Companion$lambda0(MetadataExtensions.kt)
	at kotlin.reflect.jvm.internal.impl.km.internal.extensions.MetadataExtensions$Companion$$Lambda$0.invoke(Unknown Source)
	at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:86)
	at kotlin.reflect.jvm.internal.impl.km.internal.extensions.MetadataExtensions$Companion.getINSTANCES$kotlin_metadata(MetadataExtensions.kt:92)
	at kotlin.reflect.jvm.internal.impl.km.KmClass.<init>(Nodes.kt:135)
	at kotlin.reflect.jvm.internal.impl.km.internal.ReadersKt.toKmClass(Readers.kt:56)
	at kotlin.reflect.jvm.internal.impl.km.internal.ReadersKt.toKmClass$default(Readers.kt:50)
	at kotlin.reflect.jvm.internal.KClassImpl$Data.kmClass_delegate$lambda$1(KClassImpl.kt:63)
	at kotlin.reflect.jvm.internal.KClassImpl$Data.accessor$KClassImpl$Data$lambda0(KClassImpl.kt)
	at kotlin.reflect.jvm.internal.KClassImpl$Data$$Lambda$0.invoke(Unknown Source)
	at kotlin.SafePublicationLazyImpl.getValue(LazyJVM.kt:125)
	at kotlin.reflect.jvm.internal.KClassImpl$Data.getKmClass(KClassImpl.kt:61)
	at kotlin.reflect.jvm.internal.KClassImpl.getKmClass(KClassImpl.kt:228)
	at kotlin.reflect.jvm.internal.KClassImpl.isValue(KClassImpl.kt:346)
	at com.fasterxml.jackson.module.kotlin.ReflectionCache.getValueClassReturnType(ReflectionCache.kt:109)
	at com.fasterxml.jackson.module.kotlin.ReflectionCache.findValueClassReturnType(ReflectionCache.kt:118)
	at com.fasterxml.jackson.module.kotlin.KotlinAnnotationIntrospector.findValueClassReturnType(KotlinAnnotationIntrospector.kt:162)
	at com.fasterxml.jackson.module.kotlin.KotlinAnnotationIntrospector.findNullSerializer(KotlinAnnotationIntrospector.kt:97)
	at com.fasterxml.jackson.module.kotlin.KotlinAnnotationIntrospector.findNullSerializer(KotlinAnnotationIntrospector.kt:28)
	at com.fasterxml.jackson.databind.introspect.AnnotationIntrospectorPair.findNullSerializer(AnnotationIntrospectorPair.java:354)
	at com.fasterxml.jackson.databind.introspect.AnnotationIntrospectorPair.findNullSerializer(AnnotationIntrospectorPair.java:354)
	at com.fasterxml.jackson.databind.introspect.AnnotationIntrospectorPair.findNullSerializer(AnnotationIntrospectorPair.java:358)
	at com.fasterxml.jackson.databind.ser.PropertyBuilder.buildWriter(PropertyBuilder.java:237)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory._constructWriter(BeanSerializerFactory.java:906)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.findBeanProperties(BeanSerializerFactory.java:663)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.constructBeanOrAddOnSerializer(BeanSerializerFactory.java:405)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.findBeanOrAddOnSerializer(BeanSerializerFactory.java:292)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory._createSerializer2(BeanSerializerFactory.java:237)
	at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.createSerializer(BeanSerializerFactory.java:171)
	at com.fasterxml.jackson.databind.SerializerProvider._createUntypedSerializer(SerializerProvider.java:1554)
	at com.fasterxml.jackson.databind.SerializerProvider._createAndCacheUntypedSerializer(SerializerProvider.java:1502)
	at com.fasterxml.jackson.databind.SerializerProvider.findValueSerializer(SerializerProvider.java:586)
	at com.fasterxml.jackson.databind.SerializerProvider.findTypedValueSerializer(SerializerProvider.java:869)
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:331)
	at com.fasterxml.jackson.databind.ObjectMapper._writeValueAndClose(ObjectMapper.java:4859)
	at com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:4038)
	at io.quarkus.kafka.client.serialization.ObjectMapperSerializer.serialize(ObjectMapperSerializer.java:44)
	at org.apache.kafka.common.serialization.Serializer.serialize(Serializer.java:62)
	at io.smallrye.reactive.messaging.kafka.fault.SerializerWrapper.lambda$serialize$1(SerializerWrapper.java:56)
	at io.smallrye.reactive.messaging.kafka.fault.SerializerWrapper.wrapSerialize(SerializerWrapper.java:81)
	at io.smallrye.reactive.messaging.kafka.fault.SerializerWrapper.serialize(SerializerWrapper.java:56)
	at org.apache.kafka.clients.producer.KafkaProducer.doSend(KafkaProducer.java:987)
	at org.apache.kafka.clients.producer.KafkaProducer.send(KafkaProducer.java:943)
	at io.smallrye.reactive.messaging.kafka.impl.ReactiveKafkaProducer.lambda$send$4(ReactiveKafkaProducer.java:167)
	at io.smallrye.context.impl.wrappers.SlowContextualConsumer.accept(SlowContextualConsumer.java:21)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateWithEmitter.subscribe(UniCreateWithEmitter.java:22)
	... 14 more

2025-09-18 14:35:08,129 ERROR [com.wog.dev.gat.ada.age.pri.web.AgentPrimaryWebSocketAdapter] (vert.x-eventloop-thread-9) Retries exhausted : 10 attempts against 1758206126277/1758206124096 expiration
```

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

Java 21 (GraalVM native-image)

### Mandrel or GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

mvnw

### Additional information

```
@ApplicationScoped
class AgentService {

    @Inject
    @Channel("agents")
    lateinit var resources: MutinyEmitter<OnEventStoreMessage>

    fun onJoined(event: OnAgentJoined): Uni<Void> {
        return resources.send(
            OnEventStoreAgentConnected(event.agent.id.serial, event.since)
        )
    }
}
```

--- a/extensions/kotlin/deployment/src/main/java/io/quarkus/kotlin/deployment/KotlinProcessor.java
+++ b/extensions/kotlin/deployment/src/main/java/io/quarkus/kotlin/deployment/KotlinProcessor.java
@@ -13,6 +13,7 @@ import io.quarkus.deployment.builditem.nativeimage.NativeImageResourcePatternsBu
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassFinalFieldsWritablePredicateBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveHierarchyIgnoreWarningBuildItem;
+import io.quarkus.deployment.builditem.nativeimage.ServiceProviderBuildItem;
 import io.quarkus.jackson.spi.ClassPathJacksonModuleBuildItem;
 
 public class KotlinProcessor {
@@ -52,6 +53,7 @@ public class KotlinProcessor {
      */
     @BuildStep
     void registerKotlinReflection(final BuildProducer<ReflectiveClassBuildItem> reflectiveClass,
+            BuildProducer<ServiceProviderBuildItem> serviceProvider,
             BuildProducer<NativeImageResourcePatternsBuildItem> nativeResourcePatterns,
             BuildProducer<ReflectiveHierarchyIgnoreWarningBuildItem> reflectiveHierarchyIgnoreWarning) {
 
@@ -80,5 +82,9 @@ public class KotlinProcessor {
                 new ReflectiveHierarchyIgnoreWarningBuildItem(DotName.createSimple("kotlinx.serialization.KSerializer")));
         reflectiveHierarchyIgnoreWarning.produce(new ReflectiveHierarchyIgnoreWarningBuildItem(
                 DotName.createSimple("kotlinx.serialization.descriptors.SerialDescriptor")));
+
+        serviceProvider.produce(
+                new ServiceProviderBuildItem("kotlin.reflect.jvm.internal.impl.km.internal.extensions.MetadataExtensions",
+                        "kotlin.reflect.jvm.internal.impl.km.jvm.internal.JvmMetadataExtensions"));
     }
 }

